elb_log_ingestor/main.py

```diff
@@ -0,0 +1,75 @@
+import http.server
+import os
+import pathlib
+import queue
+import sys
+import threading
+
+import boto
+import elasticsearch
+
+import api_endpoint
+import elasticsearch_shipper
+import elb_log_fetcher
+import elb_log_parse
+import stats
+
+
+def start_server():
+    parser_stats = stats.ParserStats()
+    shipper_stats = stats.ShipperStats()
+
+    elasticsearch_hosts = os.environ['ELB_INGESTOR_ELASTICSEARCH_HOSTS']
+    elasticsearch_hosts = elasticsearch_hosts.split(',')
+    es_client = elasticsearch.Elasticsearch(elasticsearch_hosts, sniff_on_start=True, sniffer_timeout=60)
+    s3_client = boto.client("s3")
+    server_address = get_server_address()
+
+    logs_to_be_processed = queue.Queue()
+    logs_processed = queue.Queue()
+    records = queue.Queue()
+    bucket = os.environ['ELB_INGESTOR_BUCKET']
+    unprocessed_prefix = os.environ.get("ELB_INGESTOR_SEARCH_PREFIX", "logs/")
+    processing_prefix = os.environ.get("ELB_INGESTOR_WORKING_PREFIX", "logs-working/")
+    processed_prefix = os.environ.get("ELB_INGESTOR_DONE_PREFIX", "logs-done/")
+    start_queue_size = int(os.environ.get("ELB_INGESTOR_START_QUEUE_SIZE", 5))
+    fetcher = elb_log_fetcher.S3LogFetcher(
+        bucket,
+        s3_client,
+        unprocessed_prefix=unprocessed_prefix,
+        processing_prefix=processing_prefix,
+        processed_prefix=processed_prefix,
+        to_do=logs_to_be_processed,
+        done=logs_processed,
+        start_queue_size=start_queue_size,
+    )
+
+    parser = elb_log_parse.LogParser(
+        logs_to_be_processed, logs_processed, parser_stats
+    )
+    shipper = elasticsearch_shipper.ElasticsearchShipper(
+        es_client, records, index_pattern, shipper_stats
+    )
+    api_endpoint = api_endpoint.ApiEndpoint(parser_stats, shipper, shipper, fetcher)
+    server = http.server.HTTPServer(server_address, api_endpoint)
+
+    fetcher_thread = threading.Thread(target=fetcher.run)
+    parser_thread = threading.Thread(target=parser.run, daemon=True)
+    shipper_thread = threading.Thread(target=shipper.run, daemon=True)
+    server_thread = threading.Thread(target=server.serve_forever, daemon=True)
+
+    fetcher_thread.start()
+    parser_thread.start()
+    shipper_thread.start()
+    server_thread.start()
+
+
+def get_server_address() -> (str, int):
```

Comment:
The default port in cf is `8080`, but you should always listen on `0.0.0.0` and read the port from the `$PORT` env var when running workloads in cf.

Reply:
this is going to run outside of CF